crates/runc-shim/src/cgroup_memory.rs

```diff
@@ -0,0 +1,184 @@
+use containerd_shim::{
+    error::{Error, Result},
+    io_error, other_error,
+};
+use std::path::Path;
+use tokio::{
+    fs::{self, read_to_string, File},
+    io::AsyncReadExt,
+    sync::mpsc::{self, Receiver},
+};
+
+use nix::sys::eventfd;
+use std::os::unix::io::{AsRawFd, FromRawFd};
+
+pub async fn get_path_from_cgorup(pid: u32) -> Result<String> {
+    let proc_path = format!("/proc/{}/cgroup", pid);
+    let path_string = read_to_string(proc_path.clone()).await.map_err(io_error!(
+        e,
+        "open {}.",
+        proc_path.clone()
+    ))?;
+
+    let (_, path) = path_string
+        .lines()
+        .find(|line| line.contains("memory"))
+        .ok_or(Error::Other("Memory line not found".into()))?
+        .split_once(":memory:")
+        .ok_or(Error::Other("Failed to parse memory line".into()))?;
+
+    Ok(path.to_string())
+}
+
+pub async fn get_existing_cgroup_mem_path(pid_path: String) -> Result<(String, String)> {
+    let (mut mount_root, mount_point) = get_path_from_mountinfo().await?;
+    if mount_root == "/" {
+        mount_root = String::from("");
+    }
+    let mount_root = pid_path.trim_start_matches(&mount_root).to_string();
+    Ok((mount_root, mount_point))
+}
+
+async fn get_path_from_mountinfo() -> Result<(String, String)> {
+    let mountinfo_path = "/proc/self/mountinfo".to_string();
+    let mountinfo_string = read_to_string(mountinfo_path.clone())
+        .await
+        .map_err(io_error!(e, "open {}.", mountinfo_path.clone()))?;
+
+    let line = mountinfo_string
+        .lines()
+        .find(|line| line.contains("cgroup") && line.contains("memory"))
+        .ok_or(Error::Other(
+            "Lines containers cgroup and memory not found in mountinfo".into(),
+        ))?;
+
+    parse_memory_mountroot(line)
+}
+
+fn parse_memory_mountroot(line: &str) -> Result<(String, String)> {
+    let mut columns = line.split_whitespace();
+    columns.nth(2);
+    let mount_root = columns.next().ok_or(Error::Other(
```

Comment:
why do we need `nth(2)` and `next` ? why not just `nth(3)` ?

Reply:
Or i can modify here as follows:
```rust
    let mount_root = columns.nth(3).ok_or(Error::Other(
        "Invalid input information about mountinfo".into(),
    ))?;
    let mount_point = columns.next().ok_or(Error::Other(
        "Invalid input information about mountinfo".into(),
    ))?;
```